### ISSUE TITLE
How can we implement loop node like Unreal Engine

### ISSUE THREAD

**Issue by MutahirMehmood** [retejs/rete#243]
Hi!
Please see the image
![image](https://user-images.githubusercontent.com/1832130/51674096-93471780-1ff1-11e9-8d4f-b9d17b490146.png)

Is there a way to implement loop nodes like unreal engine blueprints and worker somehow gets executed for each element in array?

Thanks!

**Comment by Ni55aN** [comment]
Task plugin allows to create such flow. There is a two types of sockets: actions (a white triangles) and data (circles) https://codepen.io/Ni55aN/pen/MOYPEz

**Comment by MutahirMehmood** [comment]
@Ni55aN  thanks for quick response. is there a documentation about Tasks plugin about how we can use it. like if we need to access data from an input when running task or something like that. I tried to search for documentation but couldnt find much. Thanks!

**Comment by Ni55aN** [comment]
@MutahirMehmood there is no complete documentation, but you can investigate [this example](https://codepen.io/Ni55aN/pen/rpOKNb) and [tiny code chunks](https://rete.js.org/#/docs/plugins/task)

**Comment by MutahirMehmood** [comment]
Thanks! that was helpful, I was reading Task Plugin source code to better understand how it works. So what I understood is that actual Component Worker gets replaced with a new function and the original one is call when we call Task.Run is that correct? and we can get inputs of the node in init function using task paramter using task.input is that correct as well? if you can tell more for better understanding it would be really helpful. Thanks again :)

**Comment by Ni55aN** [comment]
>  Component Worker gets replaced with a new function and the original one is called when we call Task.Run is that correct?

exactly

> get inputs of the node in init function

`task.inputs` contains input tasks and used by plugin itself. It is not recommended to use outside.
You can get input data inside worker

```js
  worker(node, inputs) {
    /// inputs["text"][0]
  }
```

**Comment by MutahirMehmood** [comment]
Right, but for example if I need to access Inputs in init function and reason I can tell you by example 

`init(task){
   let inputArray = task.inputs['someinput'][0];
   for(i = 0; i < inputArray.lenght; i++){
      task.run([inputArray[i]);
      task.reset();
   }
}`

this way worker would send each element in array to output. or is there a better way to achieve this?

**Comment by Ni55aN** [comment]
`inputArray` presents meta data for scheme processing, not runtime data
![screenshot from 2019-01-25 14-56-34](https://user-images.githubusercontent.com/8259641/51749378-df5c8f80-20b7-11e9-9a56-65c55c0572f1.png)

All operations related to the transfer of data between nodes must be performed inside the worker

**Comment by MutahirMehmood** [comment]
yes that is correct but problem is that looping through elements of Array cannot cause worker to execute again and again we need to call **task.run** each time to execute worker and send data to output. thats why I am trying to access node data inside the **init** function. Or is there any other way to achieve this?

**Comment by Ni55aN** [comment]
can you share your example on jsfiddle or smth else?

In short, inside an `init` function you can save the `task` instance to store and take it inside `worker()`

**Comment by Ni55aN** [comment]
![sketch1548429366029](https://user-images.githubusercontent.com/8259641/51754449-0077ad00-20c5-11e9-9727-866d51b4eb7e.png)

**Comment by MutahirMehmood** [comment]
I tried to create this as you suggested but for some reason my code executes very randomly and I am unable to figure out whats happening. I have created a Pen so if you could take a look please. Actually my loop gets executed once or twice or thrice but not more than that and there are 100 items in Array I get from Api call.

[https://codepen.io/mutahir/pen/ZwQqOX](url)

You would have to click **Process** button on the top to get flow executed. and you can see the output in console. Also I did add `this.closed = ['output']` but it halts the process and next node doesn't get executed, So I omitted it. 

Ideally I wanted the **ForEach** connected directly to **Fetch Data** component but it was causing Api calls every time loop iterates So, I had to create additional component (**Get Data**) that gets data from memory which was set by **Fetch Data** component. could you please check in the pen what could be wrong here.

Thanks!

**Comment by Ni55aN** [comment]
@MutahirMehmood I'll try to fork and modify it

**Comment by MutahirMehmood** [comment]
**UPDATE**

@Ni55aN  If I put `task.run(index)` in a `setTimeout` even with 0 MS timeout. it starts executing perfectly and goes through all the items in the array. New worker looks like this.

``` javascript
          worker(node, inputs, index = 0) {
                var inpArray = inputs["array"][0];
                
                if (index < inpArray.length) {
                    let el = inpArray[index];

                    index++;
                    setTimeout(() => {
                        eachTask.run(index);
                    }, 0);

                    return { element: el }
                } else {
                    eachTask.reset();
                    this.closed = ['act'];
                }
            }
```

I am trying to understand if its happing for a reason or a bug or something like that. And not sure how would it affect the performance. Let me know your thoughts please. Thanks!

**Comment by Ni55aN** [comment]
I'm working on this now, if we need  a parallel processing, the tasks must be cloned

**Comment by MutahirMehmood** [comment]
I am still examining it but what I am seeing from console output every element in array is being printed 2 times for some reason. trying to find why its doing that how ever code makes perfect sense

**Comment by Ni55aN** [comment]
@MutahirMehmood ForEach and LOG messages are displaying once for every element

**Comment by Ni55aN** [comment]
Cool!

**Comment by MutahirMehmood** [comment]
You mean multiple ForEach nodes in one graph?

**Comment by MutahirMehmood** [comment]
One more thing I removed `clone()` before `run()` and then tested and it didn't make any difference. What purpose `Clone()` is serving here

**Comment by Ni55aN** [comment]
@MutahirMehmood without `clone()` the task of Log node will refers to latest eachTask ran

How to reproduce: 
- delete `await (new Promise((res, rej) => {setTimeout(res, sec)}));`
